addons/jest/src/components/Message.tsx

```diff
@@ -0,0 +1,208 @@
+  fontWeight: 500,
+});
+const Negative = styled.strong({
+  color: colors.error,
+  fontWeight: 500,
+});
+
+interface StackTraceProps {
+  trace: MsgElement[];
+  className?: string;
+}
+
+const StackTrace = styled(({ trace, className }: StackTraceProps) => (
+  <details className={className}>
+    <summary>Callstack</summary>
+    {trace
+      .join('')
+      .trim()
+      .split(/\n/)
+      .map((traceLine, traceLineIndex) => (
+        <div key={traceLineIndex}>{traceLine.trim()}</div>
+      ))}
+  </details>
+))({
+  background: 'silver',
+  padding: 10,
+  overflow: 'auto',
+});
+
+const Main = styled(({ msg, className }) => <section className={className}>{msg}</section>)({
+  padding: 10,
+  borderBottom: '1px solid silver',
+});
+
+interface SubProps {
+  msg: MsgElement[];
+  className?: string;
+}
+
+const Sub = styled(({ msg, className }: SubProps) => (
+  <section className={className}>
+    {msg
+      .filter(item => typeof item !== 'string' || item.trim() !== '')
+      .map((item, index, list) => {
+        if (typeof item === 'string') {
+          if (index === 0 && index === list.length - 1) {
+            return item.trim();
+          }
+          if (index === 0) {
+            return item.replace(/^[\s\n]*/, '');
+          }
+          if (index === list.length - 1) {
+            return item.replace(/[\s\n]*$/, '');
+          }
+        }
+        return item;
+      })}
+  </section>
+))({
+  padding: 10,
+});
+
+interface SubgroupOptions {
+  startTrigger: (e: MsgElement) => boolean;
+  endTrigger: (e: MsgElement) => boolean;
+  grouper: (list: MsgElement[], key: number) => JSX.Element;
+  accList?: MsgElement[];
+  grouped?: MsgElement[];
+  grouperIndex?: number;
+  mode?: 'inject' | 'stop';
+  injectionPoint?: number;
+}
+
+const createSubgroup = ({
+  startTrigger,
+  endTrigger,
+  grouper,
+  accList = [],
+  grouped = [],
+  grouperIndex = 0,
+  mode,
+  injectionPoint,
+}: SubgroupOptions) => (acc: MsgElement[], item: MsgElement, i: number, list: MsgElement[]) => {
+  grouperIndex += 1;
+
+  // start or stop extraction
+  if (startTrigger(item)) {
+    mode = 'inject';
+    injectionPoint = i;
+  }
+  if (endTrigger(item)) {
+    mode = 'stop';
+  }
+
+  // push item in correct aggregator
+  if (mode === 'inject') {
+    grouped.push(item);
+  } else {
+    accList.push(item);
+  }
+
+  // on last iteration inject at detected injection point, and group
+  if (i === list.length - 1) {
+    // Provide a "safety net" when Jest returns a partially recognized "group"
+    // (recognized by acc.startTrigger but acc.endTrigger was never found) and
+    // it's the only group in output for a test result. In that case, accList
+    // will be empty, so return whatever was found, even if it will be unstyled
+    // and prevent next createSubgroup calls from throwing due to empty lists.
+    accList.push(null);
+
+    return accList.reduce<MsgElement[]>((eacc, el, ei) => {
+      switch (true) {
```

Comment:
this should really just be some if/else. a switch doesn't really make sense and makes the code more complex